dmodel/dm-macros.h

```diff
@@ -0,0 +1,144 @@
+#endif
+
+/**
+ * DM_VERSION_PREV_STABLE:
+ *
+ * Evaluates to the previous stable version.
+ */
+#if (DM_MINOR_VERSION == 99)
+#define DM_VERSION_PREV_STABLE (G_ENCODE_VERSION (DM_MAJOR_VERSION + 1, 0))
+#elif (DM_MINOR_VERSION % 2)
+#define DM_VERSION_PREV_STABLE (G_ENCODE_VERSION (DM_MAJOR_VERSION, DM_MINOR_VERSION - 1))
+#else
+#define DM_VERSION_PREV_STABLE (G_ENCODE_VERSION (DM_MAJOR_VERSION, DM_MINOR_VERSION - 2))
+#endif
+
+/**
+ * DM_VERSION_MIN_REQUIRED:
+ *
+ * A pre-processor symbol that should be defined by the user prior to including
+ * the `dmodel.h` header.
+ *
+ * The value should be one of the predefined DModel version macros,
+ * for instance: %DM_VERSION_0_0, %DM_VERSION_0_2, ...
+ *
+ * This symbol defines the earliest version of DModel that a project
+ * is required to be able to compile against.
+ *
+ * If the compiler is configured to warn about the use of deprecated symbols,
+ * then using symbols that were deprecated in version %DM_VERSION_MIN_REQUIRED,
+ * or earlier versions, will emit a deprecation warning, but using functions
+ * deprecated in later versions will not.
+ *
+ * See also: %DM_VERSION_MAX_ALLOWED
+ */
+#ifndef DM_VERSION_MIN_REQUIRED
+# define DM_VERSION_MIN_REQUIRED (DM_VERSION_CUR_STABLE)
+#endif
+
+/**
+ * DM_VERSION_MAX_ALLOWED:
+ *
+ * A pre-processor symbol that should be defined by the user prior to including
+ * the `dmodel.h` header.
+ *
+ * The value should be one of the predefined DModel version macros,
+ * for instance: %DM_VERSION_0_0, %DM_VERSION_0_2, ...
+ *
+ * This symbol defines the earliest version of DModel that a project
+ * is required to be able to compile against.
+ *
+ * If the compiler is configured to warn about the use of deprecated symbols,
+ * then using symbols that were deprecated in version %DM_VERSION_MAX_ALLOWED,
+ * or later versions, will emit a deprecation warning, but using functions
+ * deprecated in later versions will not.
+ *
+ * See also: %DM_VERSION_MIN_REQUIRED
+ */
+#ifndef DM_VERSION_MAX_ALLOWED
+# if DM_VERSION_MIN_REQUIRED > DM_VERSION_PREV_STABLE
+#  define DM_VERSION_MAX_ALLOWED (DM_VERSION_MIN_REQUIRED)
+# else
+#  define DM_VERSION_MAX_ALLOWED (DM_VERSION_CUR_STABLE)
+# endif
+#endif
+
+/* sanity checks */
+#if DM_VERSION_MAX_ALLOWED < DM_VERSION_MIN_REQUIRED
+#error "DM_VERSION_MAX_ALLOWED must be >= DM_VERSION_MIN_REQUIRED"
+#endif
+#if DM_VERSION_MIN_REQUIRED < DM_VERSION_1_0
+#error "DM_VERSION_MIN_REQUIRED must be >= DM_VERSION_1_0"
+#endif
+
+/* Every new stable minor release should add a set of macros here */
+
+#if DM_VERSION_MIN_REQUIRED >= DM_VERSION_0_0
+# define DM_DEPRECATED_IN_0_0        DM_DEPRECATED
+# define DM_DEPRECATED_IN_0_0_FOR(f) DM_DEPRECATED_FOR(f)
+#else
+# define DM_DEPRECATED_IN_0_0        _DM_EXTERN
+# define DM_DEPRECATED_IN_0_0_FOR(f) _DM_EXTERN
+#endif
+
+/* unconditional: we can't have an earlier API version */
+#define DM_AVAILABLE_IN_0_0                  _DM_EXTERN
```

Comment:
At this point, I think I'd drop all of these, and just have a `DM_AVAILABLE_IN_ALL` as the marker for the API introduced from the first release. This way we get to remove the `DM_VERSION_0_0` and every other branching.

Unless, of course, you want to release a 1.0 version now, in which case having `DM_VERSION_1_0` would still make sense, but I'd still use a `DM_AVAILABLE_IN_ALL` annotation.